src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs

```diff
@@ -0,0 +1,731 @@
+                    {
+                        goto retryRequest;
+                    }
+
+                    stream = new Http3RequestStream(request, this, quicStream);
+                    _activeRequests.Add(quicStream.StreamId, stream);
+                }
+
+                return stream.SendAsync(cancellationToken);
+            }
+            catch (QuicResetException ex)
+            {
+                // This will happen if we aborted _connection somewhere.
+                Abort(ex);
+            }
+
+        retryRequest:
+            // We lost a race between GOAWAY/abort and our pool sending the request to this connection.
+            quicStream?.Dispose();
+            return Task.FromException<HttpResponseMessage>(new HttpRequestException("Server sent GOAWAY.", null, RequestRetryType.RetryOnSameOrNextProxy));
+        }
+
+        /// <summary>
+        /// Waits for MAX_STREAMS to be raised by the server.
+        /// </summary>
+        private Task WaitForAvailableRequestStreamAsync(CancellationToken cancellationToken)
+        {
+            TaskCompletionSourceWithCancellation<bool> tcs;
+
+            lock (SyncObj)
+            {
+                long remaining = _requestStreamsRemaining;
+
+                if (remaining > 0)
+                {
+                    _requestStreamsRemaining = remaining - 1;
+                    return Task.CompletedTask;
+                }
+
+                tcs = new TaskCompletionSourceWithCancellation<bool>();
+                _waitingRequests.Enqueue(tcs);
+            }
+
+            // Note: cancellation on connection shutdown is handled in CancelWaiters.
+            return tcs.WaitWithCancellationAsync(cancellationToken).AsTask();
+        }
+
+        /// <summary>
+        /// Cancels any waiting SendAsync calls.
+        /// </summary>
+        /// <remarks>Requires <see cref="SyncObj"/> to be held.</remarks>
+        private void CancelWaiters()
+        {
+            Debug.Assert(Monitor.IsEntered(SyncObj));
+
+            while (_waitingRequests.TryDequeue(out TaskCompletionSourceWithCancellation<bool> tcs))
+            {
+                tcs.TrySetException(new HttpRequestException("Server sent GOAWAY or aborted.", null, RequestRetryType.RetryOnSameOrNextProxy));
+            }
+        }
+
+        // TODO: how do we get this event?
+        private void OnMaximumStreamCountIncrease(long newMaximumStreamCount)
+        {
+            lock (SyncObj)
+            {
+                if (newMaximumStreamCount <= _maximumRequestStreams)
+                {
+                    return;
+                }
+
+                _requestStreamsRemaining += (newMaximumStreamCount - _maximumRequestStreams);
+                _maximumRequestStreams = newMaximumStreamCount;
+
+                while (_requestStreamsRemaining != 0 && _waitingRequests.TryDequeue(out TaskCompletionSourceWithCancellation<bool> tcs))
```

Comment:
I'm a bit confused by the _waitingRequests queue here. This looks like it is draining all pending calls to SendAsync. Why do we need to do this?

Reply:
This is to respond to new MAX_STREAMS events from QUIC. `_waitingRequests` are any requests waiting for a new stream to open up.

Reply:
I think you need another place where you dequeue when a request is finished. Ex: max request stream count is 100, someone makes 101 requests, so one request is waiting. When a request finishes, it needs to call TryDequeue to start the next request, correct?

Reply:
MAX_STREAMS is a cumulative maximum, not a maximum concurrent streams:
> Note that these frames (and the corresponding transport parameters) do not describe the number of streams that can be opened concurrently. The limit includes streams that have been closed as well as those that are open.
https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#section-19.11-8